### PR TITLE
Feature/89 rload fn

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

Add utility function `rload` - see issue #89 for discussion.
Test data for unit tests taken from Stan `src/test/test-models/good`.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: cmdstanpy/utils.py [hunk 1]
```diff
@@ -216,6 +216,56 @@ def rdump(path: str, data: Dict) -> None:
...
+        next_var = idx
+        var_data = ''.join(lines[start_idx:next_var]).replace('\n', '')
+        lhs, rhs = [_.strip() for _ in var_data.split('<-')]
+        lhs = lhs.replace('"','')
+        rhs = rhs.replace('L','')
+        if rhs.startswith('structure'):
+            *_, vals, dim = rhs.replace('(', ' ').replace(')', ' ').split('c')
+            vals = [float(v) for v in vals.split(',')[:-1]]
+            dim = [int(v) for v in dim.split(',')]
+            val = np.array(vals).reshape(dim[::-1]).T
```
Comment (ahartikainen) [inline cmdstanpy/utils.py]: what is going on here?

Would this equal

    np.array(vals).reshape(dim[::-1]).T 
    --> np.array(vals, order='F').reshape(dim, order='F')
    or
    --> np.array(vals).reshape(dim, order='F')
    or
    --> np.reshape(vals, dim, order='F')

File: cmdstanpy/utils.py [hunk 2]
```diff
@@ -216,6 +216,56 @@ def rdump(path: str, data: Dict) -> None:
...
+    if idx == len(lines):
+        return None
+    start_idx = idx
+    idx += 1
+    while True:
+        while idx < len(lines) and '<-' not in lines[idx]:
+            idx += 1
+        next_var = idx
+        var_data = ''.join(lines[start_idx:next_var]).replace('\n', '')
+        lhs, rhs = [_.strip() for _ in var_data.split('<-')]
```
Comment (ahartikainen) [inline cmdstanpy/utils.py]: This is fine, just a small comment: using `_` instead of `item` seems weird to me.

File: cmdstanpy/utils.py [hunk 3]
```diff
@@ -216,6 +216,56 @@ def rdump(path: str, data: Dict) -> None:
...
 
+def rload(fname: str) -> dict:
+    """Parse data and parameter variable values from an R dump format file.
+       This parser only supports the subset of R dump data as described
+       in the "Dump Data Format" section of the CmdStan manual, i.e.,
+       scalar, vector, matrix, and array data types.
+    """
+    data_dict = {}
+    with open(fname, 'r') as fp:
+        lines = fp.readlines()
```
Comment (ahartikainen) [inline cmdstanpy/utils.py]: This is fine, larger than RAM files are not really that common.

File: cmdstanpy/utils.py [hunk 4]
```diff
@@ -216,6 +216,56 @@ def rdump(path: str, data: Dict) -> None:
...
+        else:
+            try:
+                val = int(rhs)
+            except TypeError:
+                try:
+                    val = float(rhs)
+                except TypeError:
+                    raise ValueError(rhs)
+        data_dict[lhs] = val
+        if idx == len(lines):
```
Comment (ahartikainen) [inline cmdstanpy/utils.py]: should this check be after `idx += 1` ?

File: cmdstanpy/utils.py (hunk 1 above)
Comment (mitzimorris) [reply cmdstanpy/utils.py]: agreed, that's not very clear.  I added a subroutine that processes the Rdump multi-dim structure - used an ugly regex with named groups for the essential bits - array values, array dimensions.   I ran the  Rdump file in R to see what R does, then added unit tests that verify that the array values end up in the right place.   hope this helps.

File: cmdstanpy/utils.py (hunk 2 above)
Comment (mitzimorris) [reply cmdstanpy/utils.py]: agreed, fixed.

File: cmdstanpy/utils.py (hunk 4 above)
Comment (mitzimorris) [reply cmdstanpy/utils.py]: no, this logic is correct.
line 238  will have done the incrementing already.
the loop is set up  this way to make sure that the final var decl gets processed.

**Comment by mitzimorris** [comment]
I did a rewrite of the drump parsing logic and plugged it into the `read_rdump_metric` helper fn.  @ahartikainen, if you have any free cycles, please recheck.

(this feature is as much of a PITA as everything else to do with R, but I can see its utility...)

**Pull request merged**